software/contrib/egressus_melodiam.py

```diff
@@ -0,0 +1,1091 @@
+
+                self.outputVoltageFlipFlops[idx] = not self.outputVoltageFlipFlops[idx]
+
+                # Catch buffer over-runs by detecting that not all samples were used in the last cycle
+                if self.clockStep > CLOCK_DIFF_BUFFER_LEN and not self.unClockedMode:
+                    self.bufferOverrunSamples[idx] = int(
+                        self.slewBufferPosition[idx] - self.slewBufferSampleNum[idx]
+                    )
+                    if DEBUG_MODE == 3:
+                        print(
+                            f"[{self.clockStep}][{idx}] bufferOverrunSamples: {self.bufferOverrunSamples[idx]}"
+                        )
+                    self.bufferSampleOffsets[idx] = (
+                        self.bufferSampleOffsets[idx] - self.bufferOverrunSamples[idx]
+                    )
+
+                else:
+                    self.bufferSampleOffsets[idx] = 0
+
+                # Set the target number of samples for the next cycle, factoring in any previous overruns
+                if DEBUG_MODE == 3:
+                    print(
+                        f"[{self.clockStep}][{idx}] previousTarget: {self.slewBufferSampleNum[idx]}"
+                    )
+
+                # Calculate the number of samples needed until the next clock
+                self.slewBufferSampleNum[idx] = min(
+                    SLEW_BUFFER_SIZE_IN_SAMPLES,
+                    int(
+                        (
+                            (self.averageMsBetweenClocks / 1000)
+                            * self.outputDivisions[idx]
+                            * self.samplesPerSec[idx]
+                        )
+                        - self.bufferSampleOffsets[idx]
+                    ),
+                )
+                if DEBUG_MODE == 3:
+                    print(
+                        f"[{self.clockStep}][{idx}] Time(ms) between clocks: {self.averageMsBetweenClocks}"
+                    )
+                    print(
+                        f"[{self.clockStep}][{idx}] slewBufferSampleNum: {self.slewBufferSampleNum}. Max: {SLEW_BUFFER_SIZE_IN_SAMPLES}"
+                    )
+
+                if DEBUG_MODE == 3:
+                    print(
+                        f"[{self.clockStep}][{idx}] bufferSampleOffset (AFTER): {self.bufferSampleOffsets[idx]}"
+                    )
+                    print(f"[{self.clockStep}][{idx}] newTarget: {self.slewBufferSampleNum[idx]}")
+
+                # If length is one, cycle between high and low voltages (traditional LFO)
+                # Each output uses a its configured slew shape
+                if self.patternLength == 1:
+
+                    # If square transition, set next output value to be one of the voltage extremes (flipping each time)
+                    if self.outputSlewModes[idx] == 0:
+                        self.squareOutputs[idx] = self.voltageExtremes[
+                            int(self.outputVoltageFlipFlops[idx])

```

Comment:
That's a nice suggestion, thanks. I will look at refactoring those sections.

Reply:
I did some quick and dirty performance testing which compared the time to convert a bool to an int vs pulling an int from a dict - the dict method performed slightly better after it had been called a few times - thanks for the suggestion. Updated to use a dict instead of casting the bools to ints. I am gonna skip the suggestion to use cvX.on and off for now as I am happy with the current MAX_CV_VOLTAGE method for now - hope this is OK with you. I think skipping the if statement might be slightly quicker and plus I am not really outputting gates which is the primary use of cvX.on() if I understand it correctly.